src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,180 @@
+            foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
+            {
+                //Filter the response transactions
+                OracleResponse response = tx.GetAttribute<OracleResponse>();
+                if (response is null) continue;
+
+                //Remove the request from storage
+                StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+                OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+                engine.Snapshot.Storages.Delete(key);
+
+                //Remove the id from IdList
+                key = CreateStorageKey(Prefix_IdList).Add(GetUrlHash(request.Url));
+                IdList list = engine.Snapshot.Storages.GetAndChange(key).GetInteroperable<IdList>();
+                if (!list.Remove(response.Id)) throw new InvalidOperationException();
+                if (list.Count == 0) engine.Snapshot.Storages.Delete(key);
+
+                //Mint GAS for oracle nodes
+                if (nodes.Length > 0)
+                {
+                    int index = (int)(response.Id % (ulong)nodes.Length);
+                    nodes[index].GAS += OracleRequestPrice;
+                }
+            }
+            foreach (var (account, gas) in nodes)
+            {
+                if (gas.Sign > 0) GAS.Mint(engine, account, gas);
+            }
+        }
+
+        [ContractMethod(OracleRequestPrice, CallFlags.AllowModifyStates)]
+        private void Request(ApplicationEngine engine, string url, string filter, string callback, StackItem userData, long gasForRepsonse)
+        {
+            //Check the arguments
+            if (Utility.StrictUTF8.GetByteCount(url) > MaxUrlLength
+                || (filter != null && Utility.StrictUTF8.GetByteCount(filter) > MaxFilterLength)
+                || Utility.StrictUTF8.GetByteCount(callback) > MaxCallbackLength
+                || gasForRepsonse < 0_10000000)
+                throw new ArgumentException();
+
+            //Mint gas for the response
+            engine.AddGas(gasForRepsonse);
+            GAS.Mint(engine, Hash, gasForRepsonse);
+
+            //Increase the request id
+            StorageItem item_id = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_RequestId));
+            ulong id = BitConverter.ToUInt64(item_id.Value) + 1;
+            item_id.Value = BitConverter.GetBytes(id);
+
+            //Put the request to storage
+            if (engine.Snapshot.Contracts.TryGet(engine.CallingScriptHash) is null)
+                throw new InvalidOperationException();
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Request).Add(item_id.Value), new StorageItem(new OracleRequest
+            {
+                OriginalTxid = GetOriginalTxid(engine),
```

Comment:
Pass `id`?  We need to assign `id` for the response.

Reply:
`id` is in the key.

Reply:
It's not good for neo-modules to access the request storage key directly. And the return value of `GetRequests` dosen't contain `id`.

Reply:
Done.